includes/reader-activation/sync/class-woocommerce.php

```diff
@@ -40,6 +40,118 @@ public static function should_sync_order( $order ) {
 		return true;
 	}
 
+	/**
+	 * Get the order containing what we consider to be the "Current Product" for a given user.
+	 *
+	 * All the payment fields that are synced relate to this product.
+	 *
+	 * The criteria for the "Current Product" are:
+	 * 1. The most recent subscription (either regular subscriptions or recurring donations).
+	 * 2. If no active subscriptions, the most recently cancelled or expired subscription.
+	 * 3. If no subscriptions at all, the most recent one-time donation.
+	 *
+	 * @param \WC_Customer $customer Customer object.
+	 *
+	 * @return \WC_Order|false Order object or false.
+	 */
+	private static function get_current_product_order_for_sync( $customer ) {
+		if ( ! is_a( $customer, 'WC_Customer' ) ) {
+			return false;
+		}
+
+		$user_id = $customer->get_id();
+
+		// 1. The most recent subscription (either regular subscriptions or recurring donations).
+		$active_subscriptions = WooCommerce_Connection::get_active_subscriptions_for_user( $user_id );
+		if ( ! empty( $active_subscriptions ) ) {
+			return \wcs_get_subscription( reset( $active_subscriptions ) );
+		}
+
+		// 2. If no active subscriptions, the most recently cancelled or expired subscription.
+		$most_recent_cancelled_or_expired_subscription = self::get_most_recent_cancelled_or_expired_subscription( $user_id );
+		if ( $most_recent_cancelled_or_expired_subscription ) {
+			return \wcs_get_subscription( $most_recent_cancelled_or_expired_subscription );
+		}
+
+		// 3. If no subscriptions at all, the most recent one-time donation.
+		$one_time_donation_order = self::get_one_time_donation_order_for_user( $user_id );
+		if ( $one_time_donation_order ) {
+			return $one_time_donation_order;
+		}
+
+		/**
+		 * Filter the order containing what we consider to be the "Current Product" for a given user when nothing is found.
+		 *
+		 * This is used for tests to mock the return value.
+		 *
+		 * @param false $current_product_order The returned value.
+		 * @return int $user_id The user ID.
+		 */
+		return apply_filters( 'newspack_reader_activation_get_current_product_order_for_sync', false, $user_id );
+	}
+
+	/**
+	 * Get the most recent cancelled or expired subscription for a user.
+	 *
+	 * @param int $user_id User ID.
+	 *
+	 * @return ?WCS_Subscription A Subscription object or null.
+	 */
+	private static function get_most_recent_cancelled_or_expired_subscription( $user_id ) {
+		$subcriptions = array_reduce(
+			array_keys( \wcs_get_users_subscriptions( $user_id ) ),
+			function( $acc, $subscription_id ) {
+				$subscription = \wcs_get_subscription( $subscription_id );
+				if ( $subscription->has_status( [ 'on-hold', 'cancelled', 'expired' ] ) ) {
```

Comment:
Nit: to keep things consistent, should we use [this constant](https://github.com/Automattic/newspack-plugin/blob/fix/current-product-criteria/includes/reader-revenue/woocommerce/class-woocommerce-connection.php#L19)? e.g.

```suggestion
				if ( ! $subscription->has_status( WooCommerce_Connection::ACTIVE_SUBSCRIPTION_STATUSES ) ) {
```

Reply:
That was actually intentional, but open for discussion.

My idea is that maybe we don't want to consider all non-active statuses there. Because what we are looking for are former donors/subscribers.

For example, "pending" could represent someone who never actually finished a subscription, right? or even "trash" or "deleted".

btw, in some cases "on-hold" could be the status of a subscription that was never active, it just had a failed payment and never completed...

wdyt?

Reply:
These are good points! In that case maybe it'd be a good idea to create a constant for `FORMER_SUBSCRIPTION_STATUSES` or something similar so that we can more easily explain why we're looking at these statuses in particular?

Reply:
https://github.com/Automattic/newspack-plugin/pull/3416/commits/e0b2778acb012af23faadbe67b924766e3928429